Feat(extension): Enable Async Gunicorn workers for Flask and Django e… (#747)
canonical/rockcraft: docs/reference/extensions/flask-framework.rst

    The Flask extension is compatible with the ``bare``, ``ubuntu@22.04``
    and ``ubuntu@24.04`` bases.

The Flask extension supports both synchronous and asynchronous
Gunicorn workers. If you want asynchronous workers, you have to add
``gevent`` package to the ``requirements.txt`` file.
Read more

> :ref:`Using asynchronous Gunicorn workers <flask-async-gunicorn-workers>`.

Reviewer: Since it discusses dependencies, I think the second sentence should be added as an option in the _Project requirements_, section, like so:

```rst
2. The project must...

For the project to make use of asynchronous Gunicorn workers:

- The ``requirements.txt`` file must include ``gevent`` as a dependency.
```

Author: Hey @medubelko thanks for your comments? Should I move the `Gunicorn workers. If you want asynchronous workers, you have to add
``gevent`` package to the ``requirements.txt`` file.` part to _Project requirements_ section or copy it there?

> Gunicorn workers. If you want asynchronous workers, you have to add
> ``gevent`` package to the ``requirements.txt`` file.
> Read more
> :ref:`Using asynchronous Gunicorn workers <flask-async-gunicorn-workers>`.

Reviewer: ```suggestion
Gunicorn workers.
```

   ``Flask`` declared as a dependency
2. The project must include a WSGI app with the path ``app:app``. This means
   there must be an ``app.py`` file at the root of the project with the name
   of the Flask object is set to ``app``.

For the project to make use of asynchronous Gunicorn workers:

- The ``requirements.txt`` file must include ``gevent`` as a dependency.

> Read more
> :ref:`Gunicorn worker selection <flask-gunicorn-worker-selection>`.

Reviewer: ```suggestion
```

        # list required packages or slices for your flask app below.
        - libpq-dev

.. warning::
  You can only use 1 of the dependencies parts at a time.

``parts`` > ``flask-framework/async-dependencies``
=================================================================

You can use this key to specify that you want to use async gunicorn workers in
your Flask application. It also works just like
``flask-framework/dependencies``.

Just uncomment the following lines:
.. code-block:: yaml

  parts:
    flask-framework/async-dependencies:
      python-packages:
        - gunicorn[gevent]

If your project needs additional debs to run, you can add them to

> ``stage-packages`` just like it is done in ``flask-framework/dependencies``.

Reviewer 2: Same comments as for Django

        # list required packages or slices for your flask app below.
        - libpq-dev

.. _flask-async-gunicorn-workers:

> Using asynchronous Gunicorn workers
> ===================================
> 
> If you want to use asynchronous workers, you have to add ``gevent`` package to
> the ``requirements.txt`` file. Rockcraft automatically detects this and updates
> the pebble plan to use the asynchronous workers. If you have ``gevent``
> installed in your rock but decided to use ``sync`` workers instead you can use
> the ``--args`` parameter of ``docker run`` to use ``sync`` workers instead of
> the default ``gevent``:

Reviewer: I think we can restyle this section:

- Since this is a reference, we generally want to frame this more as mechanical explantions than how-to language.
- I found a few ways to tighten up the language and clarify meaning.

```suggestion
Gunicorn worker selection
=========================

If the project has gevent as a dependency, Rockcraft automatically updates the
pebble plan to spawn asynchronous Gunicorn workers.

When the project instead needs synchronous workers, you can override the worker
type by adding ``--args flask sync`` to the Docker command that launches the
rock:
```
